File: leetcode/python/000124.binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # Time complexity: O(n)
    # Space complexity: O(h) - log(n) if tree is balanced
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        max_path_sum = float("-inf")

        def dfs(node: Optional[TreeNode]) -> int:
            if node is None:
                return 0

            greatest_left_sum = max(dfs(node.left), 0)  # if negative, we just
            greatest_right_sum = max(dfs(node.right), 0)  # reset to 0 (not adding it)
            sum_if_node_is_root = node.val + greatest_left_sum + greatest_right_sum

            nonlocal max_path_sum
            max_path_sum = max(max_path_sum, sum_if_node_is_root)

            return node.val + max(greatest_left_sum, greatest_right_sum)

        dfs(root)
        return int(max_path_sum)

File: leetcode/typescript/000124.binary-tree-maximum-path-sum.ts
/**
 * Definition for a binary tree node.
 */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// Time complexity: O(n)
// Space complexity: O(h) - log(n) if tree is balanced
function maxPathSum(root: TreeNode | null): number {
  let res = -Infinity;

  function dfs(node: TreeNode | null): number {
    if (!node) return 0;

    const greatestLeftSum = Math.max(dfs(node.left), 0);
    const greatestRightSum = Math.max(dfs(node.right), 0);
    const sum = node.val + greatestLeftSum + greatestRightSum;
    res = Math.max(res, sum);

    return node.val + Math.max(greatestLeftSum, greatestRightSum);
  }

  dfs(root);
  return res;
}
